File: app/src/test/java/bj/discogsbrowser/master/MasterPresenterTest.java
package bj.discogsbrowser.master;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.List;

import bj.discogsbrowser.model.master.Master;
import bj.discogsbrowser.model.version.MasterVersion;
import bj.discogsbrowser.network.DiscogsInteractor;
import bj.discogsbrowser.utils.schedulerprovider.TestSchedulerProvider;
import io.reactivex.Single;
import io.reactivex.schedulers.TestScheduler;

import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

/**
 * Created by Josh Laird on 09/05/2017.
 */
@RunWith(MockitoJUnitRunner.class)
public class MasterPresenterTest
{
    private MasterPresenter presenter;
    @Mock MasterContract.View view;
    @Mock DiscogsInteractor discogsInteractor;
    @Mock MasterController controller;
    private TestScheduler testScheduler = new TestScheduler();
    private String id = "123";

    @Before
    public void setUp()
    {
        presenter = new MasterPresenter(discogsInteractor, controller, new TestSchedulerProvider(testScheduler));
    }

    @After
    public void tearDown()
    {
        verifyNoMoreInteractions(view, discogsInteractor, controller);
    }

    @Test
    public void getValidData_getsDetailsAndVersions()
    {
        Master master = new Master();
        Single<Master> justMaster = Single.just(master);
        List<MasterVersion> masterVersions = MasterVersionsFactory.buildMasterVersions(2);
        Single<List<MasterVersion>> justVersions = Single.just(masterVersions);

        when(discogsInteractor.fetchMasterDetails(id)).thenReturn(justMaster);
        when(discogsInteractor.fetchMasterVersions(id)).thenReturn(justVersions);

        presenter.fetchReleaseDetails(id);
        testScheduler.triggerActions();

        verify(controller).setLoading(true);
        verify(discogsInteractor).fetchMasterDetails(id);
        verify(controller).setMaster(master);
        verify(discogsInteractor).fetchMasterVersions(id);
        verify(controller).setMasterMasterVersions(masterVersions);
    }

    @Test
    public void getInvalidData_controllerSetsError()
    {
        when(discogsInteractor.fetchMasterDetails(id)).thenReturn(Single.error(new Throwable()));

        presenter.fetchReleaseDetails(id);
        testScheduler.triggerActions();

        verify(controller).setLoading(true);
        verify(discogsInteractor).fetchMasterDetails(id);
        verify(controller).setError(true);
    }
}

File: app/src/test/java/bj/vinylbrowser/master/MasterPresenterTest.kt
package bj.vinylbrowser.master

import bj.vinylbrowser.model.master.Master
import bj.vinylbrowser.network.DiscogsInteractor
import bj.vinylbrowser.utils.schedulerprovider.TestSchedulerProvider
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.verifyNoMoreInteractions
import com.nhaarman.mockito_kotlin.whenever
import io.reactivex.Single
import io.reactivex.schedulers.TestScheduler
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.junit.MockitoJUnitRunner

/**
 * Created by Josh Laird on 22/05/2017.
 */
@RunWith(MockitoJUnitRunner::class)
class MasterPresenterTest {
    lateinit var presenter: MasterPresenter
    val view: MasterContract.View = mock()
    val discogsInteractor: DiscogsInteractor = mock()
    val controller: MasterController = mock()
    val testScheduler = TestScheduler()
    val id = "123"

    @Before
    fun setUp() {
        presenter = MasterPresenter(discogsInteractor, controller, TestSchedulerProvider(testScheduler))
    }

    @After
    fun tearDown() {
        verifyNoMoreInteractions(view, discogsInteractor, controller)
    }

    @Test
    fun getValidData_getsDetailsAndVersions() {
        val master = Master(id = id)
        val justMaster = Single.just(master)
        val masterVersions = MasterVersionsFactory.buildMasterVersions(2)
        val justVersions = Single.just(masterVersions)

        whenever(discogsInteractor.fetchMasterDetails(id)).thenReturn(justMaster)
        whenever(discogsInteractor.fetchMasterVersions(id)).thenReturn(justVersions)

        presenter.fetchReleaseDetails(id)
        testScheduler.triggerActions()

        verify(controller).setLoading(true)
        verify(discogsInteractor).fetchMasterDetails(id)
        verify(controller).setMaster(master)
        verify(discogsInteractor).fetchMasterVersions(id)
        verify(controller).setMasterMasterVersions(masterVersions)
    }

    @Test
    fun getInvalidData_controllerSetsError() {
        whenever(discogsInteractor.fetchMasterDetails(id)).thenReturn(Single.error<Master>(Throwable()))

        presenter.fetchReleaseDetails(id)
        testScheduler.triggerActions()

        verify(controller).setLoading(true)
        verify(discogsInteractor).fetchMasterDetails(id)
        verify(controller).setError(true)
    }
}